device/common/include/traccc/seeding/device/impl/count_triplets.ipp

```diff
@@ -0,0 +1,164 @@
+
+namespace traccc::device {
+
+TRACCC_HOST_DEVICE
+void count_triplets(
+    const std::size_t globalIndex, const seedfinder_config& config,
+    const sp_grid_const_view& sp_view,
+    const doublet_counter_container_types::const_view doublet_counter_view,
+    const vecmem::data::vector_view<const prefix_sum_element_t>&
+        doublet_ps_view,
+    const doublet_container_view mid_bot_doublet_view,
+    const doublet_container_view mid_top_doublet_view,
+    triplet_counter_container_types::view triplet_view) {
+
+    // Create device copy of input parameters
+    vecmem::device_vector<const prefix_sum_element_t> doublet_prefix_sum(
+        doublet_ps_view);
+
+    // Check if anything needs to be done.
+    if (globalIndex >= doublet_prefix_sum.size()) {
+        return;
+    }
+
+    // Create device copy of input parameters
+    const device_doublet_container mid_bot_doublet_device(mid_bot_doublet_view);
+
+    // Get ids
+    const prefix_sum_element_t ps_idx = doublet_prefix_sum[globalIndex];
+    const unsigned int bin_idx = ps_idx.first;
+    const unsigned int item_idx = ps_idx.second;
+
+    // Get current mid bottom doublet
+    const doublet mid_bot =
+        mid_bot_doublet_device.get_items().at(bin_idx).at(item_idx);
+
+    // Create device copy of input parameters
+    const device::doublet_counter_container_types::const_device
+        doublet_counter_device(doublet_counter_view);
+    const device_doublet_container mid_top_doublet_device(mid_top_doublet_view);
+
+    // Create device copy of output parameter
+    triplet_counter_container_types::device triplet_counter(triplet_view);
+
+    // Get all spacepoints
+    const const_sp_grid_device internal_sp_device(sp_view);
+
+    // Get internal spacepoints for current bin
+    const unsigned int& num_compat_spM_per_bin =
+        doublet_counter_device.get_headers().at(bin_idx).m_nSpM;
+
+    // Header of doublet counter : number of compatible middle sp per bin
+    // Item of doublet counter : doublet counter objects per bin
+    const vecmem::device_vector<const doublet_counter> doublet_counter_per_bin =
+        doublet_counter_device.get_items().at(bin_idx);
+
+    // Header of doublet: number of mid_top doublets per bin
+    // Item of doublet: doublet objects per bin
+    const vecmem::device_vector<const doublet> mid_top_doublets_per_bin =
+        mid_top_doublet_device.get_items().at(bin_idx);
+
+    // middle spacepoint indexes
+    const unsigned int& spM_bin = mid_bot.sp1.bin_idx;
+    const unsigned int& spM_idx = mid_bot.sp1.sp_idx;
+    // middle spacepoint
+    const traccc::internal_spacepoint<traccc::spacepoint>& spM =
+        internal_sp_device.bin(spM_bin)[spM_idx];
+    // bottom spacepoint indexes
+    const unsigned int& spB_bin = mid_bot.sp2.bin_idx;
+    const unsigned int& spB_idx = mid_bot.sp2.sp_idx;
+    // bottom spacepoint
+    const traccc::internal_spacepoint<traccc::spacepoint>& spB =
+        internal_sp_device.bin(spB_bin)[spB_idx];
```

Comment:
We should get all of these into local variables as well. Definitely at least for all of the `unsigned int` variables. (The internal spacepoint variable is maybe more debatable.)